### PR TITLE
cephfs-top: self-adapt the display according the window size

### PR DESCRIPTION
**Pull request by lxbsz**
If the terminal window size to smaller than the text displaying, the
cephfs-top tool will crash with:

`cephfs-top: exception: addwstr() returned ERR`

And also we cannot change the terminal size when the tool is running.

This will allow change the window size when the cephfs-top tool is
running and will adapt the display according to the real time window
size.

Fixes: https://tracker.ceph.com/issues/50091
Signed-off-by: Xiubo Li <xiubli@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by joscollin** [comment]
Making the window size smaller and then maximize stops the refresh. Please see the resulting screen (screenshot) attached.
cephfs-top

**Comment by lxbsz** [comment]
> Making the window size smaller and then maximize stops the refresh. Please see the resulting screen (screenshot) attached.
> cephfs-top

Forgot to handle the header window case. Have fixed it.

**Comment by lxbsz** [comment]
> > Making the window size smaller and then maximize stops the refresh. Please see the resulting screen (screenshot) attached.
> 
> Forgot to handle the header window case. Have fixed it.

**Comment by vshankar** [comment]
> If the terminal window size to smaller than the text displaying, the
> cephfs-top tool will crash with:
> 
> `cephfs-top: exception: addwstr() returned ERR`
> 
> And also we cannot change the terminal size when the tool is running.
> 
> This will allow change the window size when the cephfs-top tool is
> running and will adapt the display according to the real time window
> size.

Nice :)

I'll take a look early next week.

**Comment by lxbsz** [comment]
> > If the terminal window size to smaller than the text displaying, the
> > cephfs-top tool will crash with:
> > `cephfs-top: exception: addwstr() returned ERR`
> > And also we cannot change the terminal size when the tool is running.
> > This will allow change the window size when the cephfs-top tool is
> > running and will adapt the display according to the real time window
> > size.
> 
> Nice :)
> 
> I'll take a look early next week.

Sure, thanks.

Updated it to make the cursor invisible.

**Comment by lxbsz** [comment]
jenkins test api

File: src/tools/cephfs/top/cephfs-top [hunk 1]
```diff
@@ -36,15 +36,15 @@ FS_TOP_PROG_STR = 'cephfs-top'
...
 # min refresh interval allowed
 MIN_REFRESH_INTERVAL = 0.5
 
 # metadata provided by mgr/stats
-FS_TOP_MAIN_WINDOW_COL_CLIENT_ID = "CLIENT_ID"
-FS_TOP_MAIN_WINDOW_COL_MNT_ROOT = "MOUNT_ROOT"
-FS_TOP_MAIN_WINDOW_COL_MNTPT_HOST_ADDR = "MOUNT_POINT@HOST/ADDR"
+FS_TOP_MAIN_WINDOW_COL_CLIENT_ID = "client_id"
+FS_TOP_MAIN_WINDOW_COL_MNT_ROOT = "mount_root"
+FS_TOP_MAIN_WINDOW_COL_MNTPT_HOST_ADDR = "mount_piont@host/addr"
```
Comment (vshankar) [inline src/tools/cephfs/top/cephfs-top:47]: typo: "mount_point"

**Review by vshankar — CHANGES_REQUESTED** [review]
I'l still going through the curses API used here.

(minor comment for now)

File: src/tools/cephfs/top/cephfs-top (hunk 1 above)
Comment (lxbsz) [reply src/tools/cephfs/top/cephfs-top]: Fixed it.

File: src/tools/cephfs/top/cephfs-top [hunk 2]
```diff
@@ -179,6 +180,27 @@ class FSTop(object):
             raise FSTopException(f'error in \'perf stats\' query: {out}')
         return json.loads(buf.decode('utf-8'))
 
+    def items(self, item):
+        if item == "CAP_HIT":
+            return "chit"
```
Comment (vshankar) [inline src/tools/cephfs/top/cephfs-top:185]: Should we start documenting these? We should have documented it earlier, however, now these metrics are shown in abbreviated form. Maybe in cephfs-top man page?

Comment (lxbsz) [reply src/tools/cephfs/top/cephfs-top:185]: Yeah, make sense and agree.
I will try to doc it.

**Review by batrick — APPROVED** [review]
\o/

File: doc/man/8/cephfs-top.rst [hunk 3]
```diff
@@ -36,6 +36,20 @@ Options
...
+======================
+
+::
+
+   chit: cap hit rate
+   rlat: read latency
+   wlat: write latency
+   mlat: metadata latency
+   dlease: dentry lease rate
+   ofiles: opened file number
```
Comment (vshankar) [inline doc/man/8/cephfs-top.rst:49]: nit: number of open files.

File: doc/man/8/cephfs-top.rst [hunk 4]
```diff
@@ -36,6 +36,20 @@ Options
...
+   oicaps: pinned cap number
```
Comment (vshankar) [inline doc/man/8/cephfs-top.rst:50]: ditto

File: doc/man/8/cephfs-top.rst [hunk 5]
```diff
@@ -36,6 +36,20 @@ Options
...
+   oinodes: opened inode number
```
Comment (vshankar) [inline doc/man/8/cephfs-top.rst:51]: ditto

**Review by vshankar — APPROVED** [review]
Nice work @lxbsz

File: doc/man/8/cephfs-top.rst [hunk 6]
```diff
@@ -36,6 +36,20 @@ Options
 
    Perform a selftest. This mode performs a sanity check of ``stats`` module.
 
+Descriptions of fields
+======================
+
+::
+
+   chit: cap hit rate
```
Comment (tchaikov) [inline doc/man/8/cephfs-top.rst:44]: nit, could put
```rst
.. describe:: chit

   cap hit rate
```

Comment (lxbsz) [reply doc/man/8/cephfs-top.rst]: Done.

**Comment by lxbsz** [comment]
@batrick,

Updated it and only changed the doc/.

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2021-Apr-15

**Pull request merged**